File: src/main/java/waterpower/common/block/inventory/InventorySlot.java
package waterpower.common.block.inventory;

import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.util.EnumFacing;
import waterpower.common.block.tileentity.TileEntityInventory;
import waterpower.util.StackUtil;

public class InventorySlot {
    private final TileEntityInventory base;
    public final String name;
    protected final ItemStack[] contents;
    protected final Access access;
    public final InvSide preferredSide;

    public InventorySlot(TileEntityInventory base, String name, Access access, int count) {
        this(base, name, access, count, InvSide.ANY);
    }

    public InventorySlot(TileEntityInventory base, String name, Access access, int count, InvSide preferredSide) {
        this.contents = new ItemStack[count];

        this.base = base;
        this.name = name;
        this.access = access;
        this.preferredSide = preferredSide;

        base.addInvSlot(this);
    }

    public void readFromNBT(NBTTagCompound nbtTagCompound) {
        NBTTagList contentsTag = nbtTagCompound.getTagList("Contents", 10);

        for (int i = 0; i < contentsTag.tagCount(); i++) {
            NBTTagCompound contentTag = contentsTag.getCompoundTagAt(i);

            int index = contentTag.getByte("Index") & 0xFF;
            ItemStack itemStack = ItemStack.loadItemStackFromNBT(contentTag);

            if (itemStack == null) {
                continue;
            }

            put(index, itemStack);
        }
    }

    public void writeToNBT(NBTTagCompound nbtTagCompound) {
        NBTTagList contentsTag = new NBTTagList();

        for (int i = 0; i < this.contents.length; i++) {
            if (this.contents[i] == null) {
                continue;
            }
            NBTTagCompound contentTag = new NBTTagCompound();

            contentTag.setByte("Index", (byte) i);
            this.contents[i].writeToNBT(contentTag);

            contentsTag.appendTag(contentTag);
        }

        nbtTagCompound.setTag("Contents", contentsTag);
    }

    public int size() {
        return this.contents.length;
    }

    public ItemStack get() {
        return get(0);
    }

    public ItemStack get(int index) {
        return this.contents[index];
    }

    public void put(ItemStack content) {
        put(0, content);
    }

    public void put(int index, ItemStack content) {
        this.contents[index] = content;
    }

    public void clear() {
        for (int i = 0; i < this.contents.length; i++)
            this.contents[i] = null;
    }

    public boolean accepts(ItemStack itemStack) {
        return true;
    }

    public boolean canInput() {
        return (this.access == Access.I) || (this.access == Access.IO);
    }

    public boolean canOutput() {
        return (this.access == Access.O) || (this.access == Access.IO);
    }

    public boolean isEmpty() {
        for (ItemStack itemStack : this.contents) {
            if (itemStack != null)
                return false;
        }

        return true;
    }

    public void organize() {
        for (int dstIndex = 0; dstIndex < this.contents.length - 1; dstIndex++) {
            ItemStack dst = this.contents[dstIndex];

            if ((dst != null) && (dst.stackSize >= dst.getMaxStackSize()))
                continue;
            for (int srcIndex = dstIndex + 1; srcIndex < this.contents.length; srcIndex++) {
                ItemStack src = this.contents[srcIndex];
                if (src == null)
                    continue;
                if (dst == null) {
                    this.contents[srcIndex] = null;
                    ItemStack tmp85_83 = src;
                    dst = tmp85_83;
                    this.contents[dstIndex] = tmp85_83;
                } else if (StackUtil.isStackEqual(dst, src)) {
                    int space = dst.getMaxStackSize() - dst.stackSize;

                    if (src.stackSize <= space) {
                        this.contents[srcIndex] = null;
                        dst.stackSize += src.stackSize;
                    } else {
                        src.stackSize -= space;
                        dst.stackSize += space;
                        break;
                    }
                }
            }
        }
    }

    @Override
	public String toString() {
        String ret = this.name + "[" + this.contents.length + "]: ";

        for (int i = 0; i < this.contents.length; i++) {
            ret = ret + this.contents[i];

            if (i >= this.contents.length - 1)
                continue;
            ret = ret + ", ";
        }

        return ret;
    }

    public TileEntityInventory getTileEntity() {
        return base;
    }

    public ItemStack[] getCopiedContent() {
        return StackUtil.getCopiedStacks(contents);
    }

    public boolean isEquals(ItemStack[] is) {
        return StackUtil.isStacksEqual(is, contents);
    }

    public static enum InvSide {
        ANY, TOP, BOTTOM, SIDE;

        public boolean matches(EnumFacing side) {
            return (this == ANY) || ((side == EnumFacing.DOWN) && (this == BOTTOM)) || ((side == EnumFacing.UP) && (this == TOP)) || ((side.getIndex() >= 2) && (side.getIndex() <= 5) && (this == SIDE));
        }
    }

    public static enum Access {
        NONE, I, O, IO;
    }
}

File: src/main/kotlin/waterpower/common/block/inventory/InventorySlot.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.common.block.inventory

import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.nbt.NBTTagList
import net.minecraft.util.EnumFacing
import waterpower.common.block.tile.TileEntityInventory
import waterpower.util.*
import java.util.*
import kotlin.NoSuchElementException

open class InventorySlot(val tileEntity: TileEntityInventory, val name: String, protected val access: InventorySlot.Access, count: Int, val preferredSide: InventorySlot.InvSide = InventorySlot.InvSide.ANY)
    : Iterable<ItemStack> {
    protected val contents = Array<ItemStack>(count, { emptyStack })
    var stackSizeLimit = 64

    init {
        tileEntity.addInvSlot(this)
    }

    fun readFromNBT(nbtTagCompound: NBTTagCompound) {
        val contentsTag = nbtTagCompound.getTagList("Contents", 10)

        for (i in 0..contentsTag.tagCount() - 1) {
            val contentTag = contentsTag.getCompoundTagAt(i)

            val index = contentTag.getByte("Index").toInt() and 0xFF
            val itemStack = ItemStack(contentTag)
            put(index, itemStack)
        }
    }

    fun writeToNBT(nbtTagCompound: NBTTagCompound) {
        val contentsTag = NBTTagList()

        for (i in this.contents.indices) {
            if (isStackEmpty(this.contents[i])) {
                continue
            }
            val contentTag = NBTTagCompound()

            contentTag.setByte("Index", i.toByte())
            this.contents[i].writeToNBT(contentTag)

            contentsTag.appendTag(contentTag)
        }

        nbtTagCompound.setTag("Contents", contentsTag)
    }

    fun size() = this.contents.size

    operator fun get(index: Int = 0) = this.contents[index]

    fun put(content: ItemStack) = put(0, content)

    fun put(index: Int, content: ItemStack) {
        this.contents[index] = content
    }

    fun clear() = Arrays.fill(contents, emptyStack)
    fun clear(index: Int) = put(index, emptyStack)

    open fun accepts(stack: ItemStack) = true

    fun markDirty() {}

    open fun canInput() = this.access == Access.I || this.access == Access.IO
    open fun canOutput() = this.access == Access.O || this.access == Access.IO

    fun isEmpty(): Boolean {
        for (itemStack in this.contents)
            if (!isStackEmpty(itemStack))
                return false

        return true
    }

    fun organize() {
        for (dstIndex in 0..this.contents.size - 1 - 1) {
            var dst: ItemStack = this.contents[dstIndex]

            if (!isStackEmpty(dst) && getCount(dst) >= dst.maxStackSize)
                continue
            for (srcIndex in dstIndex + 1..this.contents.size - 1) {
                var src = this.contents[srcIndex]
                if (isStackEmpty(src)) continue
                if (isStackEmpty(dst)) {
                    this.contents[srcIndex] = emptyStack
                    val tmp = src
                    dst = tmp
                    this.contents[dstIndex] = tmp
                } else if (isStackEqual(dst, src)) {
                    val space = dst.maxStackSize - getCount(dst)

                    if (getCount(src) <= space) {
                        this.contents[srcIndex] = emptyStack
                        dst = grow(dst, getCount(src))
                    } else {
                        src = shrink(src)
                        dst = grow(dst)
                        break
                    }
                }
            }
        }
    }

    override fun toString(): String {
        var ret = "$name[${this.contents.size}]: "

        for (i in this.contents.indices) {
            ret += this.contents[i]

            if (i >= this.contents.size - 1)
                continue
            ret += ", "
        }

        return ret
    }

    override fun iterator(): Iterator<ItemStack> {
        return object : Iterator<ItemStack> {
            var idx = 0

            override fun next(): ItemStack {
                if (idx >= this@InventorySlot.contents.size)
                    throw NoSuchElementException()
                else
                    return this@InventorySlot.contents[idx++]
            }

            override fun hasNext() = idx < this@InventorySlot.contents.size
        }
    }

    val copiedContent: Array<ItemStack>
        get() = getCopiedStacks(contents)


    fun isEquals(stack: Array<ItemStack>): Boolean {
        return isStacksEqual(stack, contents)
    }

    enum class InvSide(val acceptedSides: Set<EnumFacing>) {
        ANY(setOf(EnumFacing.DOWN, EnumFacing.UP, EnumFacing.NORTH, EnumFacing.SOUTH, EnumFacing.WEST, EnumFacing.EAST)),
        TOP(setOf(EnumFacing.UP)),
        BOTTOM(setOf(EnumFacing.DOWN)),
        SIDE(setOf(EnumFacing.NORTH, EnumFacing.SOUTH, EnumFacing.WEST, EnumFacing.EAST)),
        NOTSIDE(emptySet());

        fun matches(side: EnumFacing) = acceptedSides.contains(side)
    }

    enum class Access {
        NONE, I, O, IO;

        fun isInput() = this == I || this == IO

        fun isOutput() = this == O || this == IO
    }
}